File: stage-common/src/main/java/io/github/solusmods/eternalcore/stage/api/StageEvents.java
package io.github.solusmods.eternalcore.stage.api;

import dev.architectury.event.Event;
import dev.architectury.event.EventFactory;
import dev.architectury.event.EventResult;
import io.github.solusmods.eternalcore.network.api.util.Changeable;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.world.entity.LivingEntity;

public interface StageEvents {
    Event<SetStageEvent> SET_STAGE = EventFactory.createEventResult();
    Event<StageTickEvent> STAGE_PRE_TICK = EventFactory.createLoop();
    Event<StageTickEvent> STAGE_POST_TICK = EventFactory.createLoop();
    Event<TrackStageEvent> TRACK_STAGE = EventFactory.createEventResult();
    Event<ReachStageEvent> REACH_STAGE = EventFactory.createEventResult();

    @FunctionalInterface
    interface SetStageEvent {
        EventResult set(StageInstance instance, LivingEntity owner, StageInstance newInstance, boolean advancement, Changeable<Boolean> notifyPlayer, Changeable<MutableComponent> stageMessage);
    }

    @FunctionalInterface
    interface StageTickEvent {
        void tick(StageInstance instance, LivingEntity owner);
    }

    @FunctionalInterface
    interface TrackStageEvent {
        EventResult track(StageInstance instance, LivingEntity owner, boolean advancement, Changeable<Boolean> notifyPlayer, Changeable<MutableComponent> stageMessage);
    }

    @FunctionalInterface
    interface ReachStageEvent {
        EventResult reach(StageInstance instance, LivingEntity owner, boolean advancement, Changeable<Boolean> notifyPlayer, Changeable<MutableComponent> stageMessage);
    }
}

File: stage-common/src/main/kotlin/io/github/solusmods/eternalcore/stage/api/StageEvents.kt
package io.github.solusmods.eternalcore.stage.api

import dev.architectury.event.Event
import dev.architectury.event.EventFactory
import dev.architectury.event.EventResult
import io.github.solusmods.eternalcore.network.api.util.Changeable
import net.minecraft.network.chat.MutableComponent
import net.minecraft.world.entity.LivingEntity

interface StageEvents {
    fun interface SetStageEvent {
        fun set(
            instance: StageInstance,
            owner: LivingEntity,
            newInstance: StageInstance,
            advancement: Boolean?,
            notifyPlayer: Changeable<Boolean?>?,
            stageMessage: Changeable<MutableComponent?>?
        ): EventResult
    }

    fun interface StageTickEvent {
        fun tick(instance: StageInstance, owner: LivingEntity)
    }

    fun interface TrackStageEvent {
        fun track(
            instance: StageInstance,
            owner: LivingEntity,
            advancement: Boolean?,
            notifyPlayer: Changeable<Boolean?>?,
            stageMessage: Changeable<MutableComponent?>?
        ): EventResult
    }

    fun interface ReachStageEvent {
        fun reach(
            instance: StageInstance,
            owner: LivingEntity,
            advancement: Boolean?,
            notifyPlayer: Changeable<Boolean?>?,
            stageMessage: Changeable<MutableComponent?>?
        ): EventResult
    }

    companion object {
        val SET_STAGE: Event<SetStageEvent> = EventFactory.createEventResult<SetStageEvent>()
        val STAGE_PRE_TICK: Event<StageTickEvent> = EventFactory.createLoop<StageTickEvent>()
        val STAGE_POST_TICK: Event<StageTickEvent> = EventFactory.createLoop<StageTickEvent>()
        val TRACK_STAGE: Event<TrackStageEvent> = EventFactory.createEventResult<TrackStageEvent>()
        val REACH_STAGE: Event<ReachStageEvent> = EventFactory.createEventResult<ReachStageEvent>()
    }
}
